chti/abnormal_task_perf_analyzer.py

```diff
@@ -540,11 +540,27 @@ def marked_pdf(metrics):
     return normal_model
 
 
+def impute_zero_sparse_rows(arr):
+    def impute_zero_sparse_row(row):
+        if numpy.sum(numpy.nan_to_num(row)) > 0:
+            return row
+        return row + 1.0
+
+    return numpy.apply_along_axis(impute_zero_sparse_row, 1, arr)
+
+
 def build_feature_abnormality_df(features_df, metrics_df, normal_model):
```

Comment:
This is definitely something to look at @pld @lincmba, as part of sanity-checking the sample data I noticed that we weren't scaling the abnormality contributions of each feature (aka task) by the actual difference-from-mean.  That means that we were just figuring out which features *could* have made things abnormal but not the ones that actually were.  Intuitively this means that one abnormal task would "drag" otherwise normal tasks (for that worker) to look abnormal if there were other workers who had those tasks as abnormal too.

Reply:
There's also a second, unrelated issue I also found (need to open a ticket) - because our task features are "surprise" and we're just looking for any abnormal surprise, the analysis is also picking up cases where surprise is abnormally *low* (aka the worker matches the historical model exactly).  I'm not sure we want this considered in the results.  In the real data sets we tried this was dwarfed by any kind of surprising results, but it's easy to see in the synthetic data set when the analyzed worker task data is drawn from the exact same distribution as the historical worker task data.